src/libraries/Microsoft.Extensions.Configuration.Binder/gen/ConfigurationBinderSourceGenerator.Emitter.cs

```diff
@@ -0,0 +1,401 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Diagnostics;
+using System.Linq;
+using System.Text;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.DotnetRuntime.Extensions;
+using Microsoft.CodeAnalysis.Text;
+
+namespace Microsoft.Extensions.Configuration.SourceGeneration
+{
+    public sealed partial class ConfigurationBinderSourceGenerator
+    {
+        private sealed partial class Emitter
+        {
+            private readonly ConfigurationBinderSourceGenerationContext _sourceGenerationContext;
+            private readonly Parser.InputTypes _targetTypes;
+
+            private readonly Queue<ITypeSymbol> _privateBindMethodGen_QueuedTypes = new();
+#pragma warning disable RS1024
+            private readonly HashSet<ITypeSymbol> _bindMethodGen_ProcessedTypes = new(SymbolEqualityComparer.Default);
+            private readonly HashSet<ITypeSymbol> _unsupportedTypes = new(SymbolEqualityComparer.Default);
+#pragma warning restore RS1024
+
+            private readonly INamedTypeSymbol? _typeSymbolForIConfigurationSection;
+
+            // Collections
+            private readonly INamedTypeSymbol _typeSymbolForIEnumerable;
+            private readonly INamedTypeSymbol? _typeSymbolForGenericIEnumerable;
+
+            private readonly INamedTypeSymbol? _typeSymbolForGenericIDictionary;
+            private readonly INamedTypeSymbol? _typeSymbolForGenericDictionary;
+
+            // Postfix for stringValueX variables used to save config value indexer
+            // results e.g. if (configuration["Key"] is string stringValue0) { ... }
+            private int _stringValueCount;
+
+            private readonly SourceWriter _writer = new();
+
+            public Emitter(
+                ConfigurationBinderSourceGenerationContext sourceGenerationContext, Parser.InputTypes targetTypes)
+            {
+                _sourceGenerationContext = sourceGenerationContext;
+                _targetTypes = targetTypes;
+
+                Compilation compilation = _sourceGenerationContext.Compilation;
```

Comment:
I would **strongly** discourage taking the entire `Compilation` object as a dependency in the emitter, as this is precisely what is creating the performance issues with incremental compilation we're seeing in the STJ source generator:

https://github.com/dotnet/runtime/issues/68353

This is called out as an anti-pattern in the [incremental compilation guidelines](https://github.com/dotnet/roslyn/blob/main/docs/features/incremental-generators.md#authoring-a-cache-friendly-generator). Per our earlier private conversation, this is another reason why we should be adopting a pipeline-oriented approach to designing the source generator: one phase for extracting the source gen model from Roslyn data and a distinct phase for emitting source code.

Reply:
👍 

> we should be adopting a pipeline-oriented approach to designing the source generator: one phase for extracting the source gen model from Roslyn data and a distinct phase for emitting source code

Yeah this is what the JSON generator implementation strived to do as well, leading to the creation of `TypeGenerationSpec` et al.

I'll refactor to do the same here (prob in this PR), and that might present an opportunity for a shared abstraction. Furthermore, I believe `TypeGenerationSpec` should have more specific derivatives like `DictionaryGenerationSpec`, `Enumerable...`, `Object...` etc.

Reply:
Awesome! I would recommend using C# records instead of regular classes when defining the models since they give you structural equality for free. This would let you define an `IncrementalValueProvider<SourceGenerationSpec>` that Roslyn can cache based on the equality of `SourceGenerationSpec` before it actually calls into the emitter phase.

Reply:
@eiriktsarpalis how does https://github.com/dotnet/runtime/pull/82179/commits/32be0a4e6a8549ad6fb6c4ae530381b0187647ef look as a first step toward leveraging the incremental feature?

If it's okay I think we can introduce some caching of `TypeSpec`s and emitted code for each type if it's impactful. This would be in a follow up PR.

Reply:
@eiriktsarpalis how does https://github.com/dotnet/runtime/pull/82179/commits/32be0a4e6a8549ad6fb6c4ae530381b0187647ef look as a first step toward leveraging the incremental feature?

If it's okay I think we can introduce some caching of `TypeSpec`s and emitted code for each type if it's impactful. This would be in a follow up PR.

Reply:
https://github.com/dotnet/runtime/pull/82179#discussion_r1131394794.